### ISSUE TITLE
go vet CI check is broken for Go >= 1.12

### ISSUE THREAD

**Issue by ire-and-curses** [stellar/go#1038]
While working through #1030, I discovered a few issues with our CI usage of `go vet`.

1. Shadow flag was removed in 1.12. See [release notes here](https://golang.org/doc/go1.12#vet) and [issue here](https://github.com/golang/go/issues/29260). Now we need to build and install a separate shadow analyzer, before running the (modified) comand.
2. There's an issue with the script where shadowed vars are not triggering a failure from the script, even on earlier versions of Go. I repro'd this locally in 1.10. Sample output:
```
clients/horizon/main_test.go:80: declaration of "err" shadows declaration at clients/horizon/main_test.go:73
exp/clients/horizon/main_test.go:253: declaration of "err" shadows declaration at exp/clients/horizon/main_test.go:218
services/horizon/internal/init.go:161: declaration of "err" shadows declaration at services/horizon/internal/init.go:151
```

Guessing the output may be going to `stderr` instead of `stdout`.
3. The invocation syntax for Go 1.12+ has changed (drop `tool` command). However even with changed syntax this was failing for me in Travis. To debug will need to install Go 1.12 locally.

Creating this issue because I've run out of time to debug this further.

**Comment by tomquisel** [comment]
@bartekn @howardtw What are your feelings on shadowed variables? Is it important to check or should we turn this off?

**Comment by bartekn** [comment]
I think we should leave checking for shadowed variables if possible. I introduced a bug in the past because the variable was shadowed: https://github.com/stellar/backend/commit/8209d864b26946116ae83b2d101ce9f857424b07. More [explanations](https://stellarfoundation.slack.com/archives/C02B04RMK/p1513174937000539).

**Comment by maria1118** [comment]
Yes I need help getting the invoice number please

**Comment by tomquisel** [comment]
@bartekn is this resolved with your new static analysis changes?

**Comment by howardtw** [comment]
Maybe we should keep it given that @bartekn encountered a bug that could have been caught. Go 1.12 dropping the experimental flag doesn’t mean it’s not helpful 😃 I’ll have to take a closer look at the reason why they drop it as well as the static analysis tho.

**Comment by bartekn** [comment]
> @bartekn is this resolved with your new static analysis changes?

It looks like staticcheck doesn't [check](https://staticcheck.io/docs/checks) for shadowed variables.

**Comment by leighmcculloch** [comment]
I need to fix this for #1700. I've swapped it out with the new `shadow` tool that hooks into `go vet` via `-vettool`. In the process of fixing it I've now got failures, namely I see all these errors now:

```
# github.com/stellar/go/clients/horizon
clients/horizon/main_test.go:79:12: declaration of "err" shadows declaration at line 73
# github.com/stellar/go/clients/horizonclient
clients/horizonclient/main_test.go:422:12: declaration of "err" shadows declaration at line 388
# github.com/stellar/go/exp/ingest/processors
exp/ingest/processors/csv_printer.go:66:8: declaration of "err" shadows declaration at line 43
exp/ingest/processors/csv_printer.go:122:8: declaration of "err" shadows declaration at line 43
# github.com/stellar/go/services/bridge/internal/listener
services/bridge/internal/listener/payment_listener_test.go:529:9: declaration of "err" shadows declaration at line 521
# github.com/stellar/go/services/bridge/internal/handlers
services/bridge/internal/handlers/request_handler_builder.go:98:7: declaration of "err" shadows declaration at line 23
services/bridge/internal/handlers/request_handler_payment.go:194:7: declaration of "err" shadows declaration at line 163
# github.com/stellar/go/services/horizon/internal/db2/history
services/horizon/internal/db2/history/operation.go:93:6: declaration of "err" shadows declaration at line 86
# github.com/stellar/go/services/horizon/internal           
services/horizon/internal/init.go:197:7: declaration of "err" shadows declaration at line 187
# github.com/stellar/go/services/ticker/internal                        
services/ticker/internal/actions_trade.go:82:13: declaration of "err" shadows declaration at line 74
# github.com/stellar/go/tools/stellar-sign                             
tools/stellar-sign/main.go:109:8: declaration of "err" shadows declaration at line 106
```

Unless anyone is already working on these, I'll fix these in #1700.

**Comment by leighmcculloch** [comment]
If you take a look in #1700 I've fixed the places that were shadowed, they were all `err`s. After looking at all of them I'm not convinced that preventing these shadow instances is in our best interest. Shadowing errors is common place in Go code, and choosing not to shadow can cause errors because developers are expecting that any error will be named `err` and your inner error might not be, which can cause folks to accidentally return outer errors.

@bartekn I'd value your thoughts on whether we should keep the shadow check given that it's going to require us to not shadow err's.

**Comment by ire-and-curses** [comment]
Yeah it's a fair point. I think I'm in favour of the shadow checks though. Several of the ones that came up during SDK development were real bugs which I was glad to have caught. And to be honest I'm not sure that having the inner scope is particularly helpful most of the time.

Here's an example of the kind of thing I'm worried about: https://github.com/golang/go/issues/21114

More discussion here: https://github.com/golang/go/issues/377

**Comment by bartekn** [comment]
> @bartekn I'd value your thoughts on whether we should keep the shadow check given that it's going to require us to not shadow err's.

I agree with @ire-and-curses. We need shadow checks but maybe not necessarily for tests. I can see there are a few instances that were not detected previously (ex. in `csv_printer.go`). I'm wondering why.

**Comment by leighmcculloch** [comment]
I'll keep the changes. I don't see anyway to limit `shadow` to non test code unfortunately. The `go vet` tool supports `-tests=false`, but not in conjunction with `-vettool`. I'll merge the fixes with #1700 then and run with it. I'll close this issue when it merges.
